Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,153 @@
+import enum
+
+from types import TracebackType
+from typing import final, Optional, Type
+
+from . import events
+from . import exceptions
+from . import tasks
+
+
+__all__ = (
+    "Timeout",
+    "timeout",
+    "timeout_at",
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    EXPIRING = "expiring"
+    EXPIRED = "expired"
+    EXITED = "finished"
+
+
+@final
+class Timeout:
+
+    def __init__(self, when: Optional[float]) -> None:
+        self._state = _State.CREATED
+
+        self._timeout_handler: Optional[events.TimerHandle] = None
+        self._task: Optional[tasks.Task] = None
+        self._when = when
+
+    def when(self) -> Optional[float]:
+        return self._when
+
+    def reschedule(self, when: Optional[float]) -> None:
+        assert self._state is not _State.CREATED
+        if self._state is not _State.ENTERED:
+            raise RuntimeError(
+                f"Cannot change state of {self._state.value} Timeout",
+            )
+
+        self._when = when
+
+        if self._timeout_handler is not None:
+            self._timeout_handler.cancel()
+
+        if when is None:
+            self._timeout_handler = None
+        else:
+            loop = events.get_running_loop()
+            self._timeout_handler = loop.call_at(
+                when,
+                self._on_timeout,
+            )
+
+    def expired(self) -> bool:
+        """Is timeout expired during execution?"""
+        return self._state in (_State.EXPIRING, _State.EXPIRED)
+
+    def __repr__(self) -> str:
+        info = ['']
+        if self._state is _State.ENTERED:
+            when = round(self._when, 3) if self._when is not None else None
+            info.append(f"when={when}")
+        info_str = ' '.join(info)
+        return f"<Timeout [{self._state.value}]{info_str}>"
+
+    async def __aenter__(self) -> "Timeout":
+        self._state = _State.ENTERED
+        self._task = tasks.current_task()
+        if self._task is None:
+            raise RuntimeError("Timeout should be used inside a task")
+        self.reschedule(self._when)
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> Optional[bool]:
+        assert self._state in (_State.ENTERED, _State.EXPIRING)
+
+        if self._timeout_handler is not None:
+            self._timeout_handler.cancel()
+            self._timeout_handler = None
+
+        if self._state is _State.EXPIRING:
+            self._state = _State.EXPIRED
+
+            if (self._task.uncancel() == 0
+                and exc_type in (None, exceptions.CancelledError)
```

Comment:
Agree, the `CancelledError` check is required to don't swallow exceptions raised by inner `finally` blocks.
But, if we go this way, the check for `None` should be dropped also.
If the inner code suppresses CancelledError -- `timeout()` context manager should not raise `TimeoutError` even if the timeout expired, isn't it?

Reply:
The current behavior (TimeoutError is raised even if no *inner await* is cancelled) looks like a partial level-triggering.  By *partial* I mean that we raise only on `__aexit__` but don't terminate other internal long-running code that caught CancelledError and performs time-consuming cleanup. The level-triggered cancellation is out of the issue scope, sure.  I doesn't look like an easy and obvious thing.

The following example is trivial:
```
async with timeout(1):
    time.sleep(2)  # Pretend to do 2 seconds of CPU work
    await asyncio.sleep(1)  <<< cancelled here
```
Timeout initiates the current task cancellation, it breaks `await sleep()` and everything goes as expected.
If the timeout doesn't cancel the task -- no TimeoutError is raised, `await` is required.

This rule better fits my mind.

Failed tests are similar, and they suppress CancelledError:
```
    async def test_waiter_cancelled(self):
        with self.assertRaises(TimeoutError):
            async with asyncio.timeout(0.01):
                with self.assertRaises(asyncio.CancelledError):
                    await asyncio.sleep(10)
```

`with self.assertRaises(asyncio.CancelledError)` swallows cancellation. TimeoutError should not be raised. It fits edge-triggered behavior. The tidy code should re-raise CancelledError instead of suppressing it.
If you are agree, I can fix tests.

Reply:
Okay, so the test failures are expected and indicate buggy tests (suppressing the CancelledError instead of reraising was a mistake).

I agree with fixing the tests, and removing the None check. Please go ahead!

Reply:
Done.
I'll re-run buildbots several times to make sure that delays and timeouts are big enough to pass on slow test boxes.